### PR TITLE
Add `noqa` to the keywords to start ignoring bears

### LINKED ISSUE #2533: Add `noqa` to the keywords to start ignoring bears
Some other linters use that so this will make coala backwards compatible

### PR DESCRIPTION
**Pull request by RohanVB**
Some other linters use 'noqa' so the addition of this keyword would make coala backwards compatible.
closes #2533

### CODE REVIEW

**Comment by aptrishu** [comment]
The two commits do the same thing ?

**Comment by RohanVB** [comment]
No, the first commit removed a vital block of code which was added in the second commit. @aptrishu

**Comment by aptrishu** [comment]
Exactly, so squash them into a single commit. There is no need for an extra commit. It'd be easier to follow then.

**Comment by aptrishu** [comment]
Also update the body message in your PR.

**Comment by sils** [comment]
is this a duplicate of your other PR? Please close your other PR then if
it's no longer needed.

**Comment by aptrishu** [comment]
Please don't delete comments on Github. @RohanVB 
It makes other comments look like meaningless.

**Comment by sils** [comment]
@RohanVB wanna set your author correctly? Or should we ignore it?

**Comment by sils** [comment]
the code looks good on first sight, I'd like to see at least one testcase for it

**Comment by sils** [comment]
maybe we can improve on the variable names :/

**Comment by RohanVB** [comment]
Hey @sils I did `git config user.email
rohan.vbh@gmail.com
` - Did that update the author for the commit on https://github.com/coala/coala/pull/2883?
the other PR : https://github.com/coala/coala/pull/2877 is closed.

**Comment by RohanVB** [comment]
Test case: 
```
vim testme.py

code = "that's checked normally"

# Start noqa LineLengthBear
unwrappable_string_2 = unwrappable_string + "yeah it goes even further..."
another_unwrappable_string = unwrappable_string + unwrappable_string_2
# Stop noqa

# noqa LineLengthBear
unwrappable_string = "some string that is long and would exceed the limit"
```

```
$ coala
[DEBUG][08:58:12] Registered VCS backend: git
[DEBUG][08:58:13] Registered VCS backend: hg
[DEBUG][08:58:13] Registered VCS backend: svn
[DEBUG][08:58:13] Registered VCS backend: bzr
Executing section Default...
Executing section python...
Executing section autopep8...
Executing section linelength...
Executing section DOCS...
Executing section commit...

```
Should I make another PR with a test in ProcessingTest.py aswell?

Also, I'm unsure of what a better variable name could be here because I put the variables to start/stop ignoring under appropriate names. I'm unsure on what the 'keywords' and 'list_ignore' could be though.

**Comment by sils** [comment]
the git config only sets it for future commits, you have to do commit --amend --reset-author or something (use google, it's simple ;))

**Comment by RohanVB** [comment]
Updated the author. Hopefully that worked.

File: coalib/processes/Processing.py [hunk 1]
```diff
@@ -437,13 +437,18 @@ def yield_ignore_ranges(file_dict):
         stop_ignoring = False
         for line_number, line in enumerate(file, start=1):
             # Before lowering all lines ever read, first look for the biggest
-            # common substring, case sensitive: I*gnor*e, start i*gnor*ing.
-            if 'gnor' in line:
+            # common substring, case sensitive: I*gnor*e, start i*gnor*ing
+            # and for N*oqa start n*oqa
```
Comment (adtac) [inline coalib/processes/Processing.py]: I don't think anyone uses `start noqa`.

Also the `*`s in the `I*gnor*e` signify the substring you're searching for. You start a `*` with `N*oqa`, but it's never ended.

So this should just be `case sensitive: I*gnor*e, start i*gnor*ing, N*oqa*.`

File: coalib/processes/Processing.py [hunk 2]
```diff
@@ -437,13 +437,18 @@ def yield_ignore_ranges(file_dict):
...
+            stop_ignore = ("stop ignoring", "stop noqa")
+            list_ignore = ("ignore ", "noqa ")
+            if any(s in line for s in keywords):
                 line = line.lower()
-                if "start ignoring " in line:
+                if any(s1 in line for s1 in start_ignore):
                     start = line_number
                     bears = get_ignore_scope(line, "start ignoring ")
-                elif "stop ignoring" in line:
+                elif any(s2 in line for s2 in stop_ignore):
```
Comment (adtac) [inline coalib/processes/Processing.py]: Uhh, I think you've got this wrongly. `noqa` doesn't function exactly like our `ignore` statements. `noqa` is used only on one line. For example:

```
    text = input()
    test=test+"hello"  # noqa
    print(test)
```

This should disable coala just for line 2.

I'm sorry, my description in the issue was lacking, my fault.

**Comment by RohanVB** [comment]
Reminder to self: Add testcases for ProcessingTest.py in the same commit.

**Comment by sils** [comment]
the git config only sets it for future commits, you have to do commit
--amend --reset-author or something (use google, it's simple ;))

Sincerely,

Lasse Schuirmann

lasse@schuirmann.net
http://coala.io/ - http://viperdev.io/ - http://gitmate.com/

File: tests/processes/ProcessingTest.py [hunk 3]
```diff
@@ -399,6 +399,17 @@ def test_yield_ignore_ranges(self):
             self.assertEqual(test_source_range.end.line, 2)
             self.assertEqual(test_source_range.end.column, 43)
 
+        test_file_dict_a = {'f':
+                            ('# noqa aBear\n',
```
Comment (RohanVB) [inline tests/processes/ProcessingTest.py]: That makes sense. I'll change it then.

Comment (adtac) [inline tests/processes/ProcessingTest.py]: yeah, `# noqa` should get applied on any bear. BTW, with this example, there's no *intuitive* way to stop ignore? using `# stop ignoring` doesn't make sense when we start with `# noqa`. I thought we agreed that `# noqa` will be the one-line equivalent for start ignoring?

but IMO we could allow for bear params; if no params then match everything

File: coalib/processes/Processing.py [hunk 4]
```diff
@@ -430,6 +431,14 @@ def get_ignore_scope(line, keyword):
         return list(StringConverter(toignore, list_delimiters=', '))
 
 
+def get_noqa_scope(line, keyword):
+    toignore = line[line.rfind(keyword) + len(keyword):]
+    if toignore.startswith('all'):
```
Comment (RohanVB) [inline coalib/processes/Processing.py]: I need some help here. `toignore.startswith(''): ` doesn't do the job; but it needs to check if there is nothing after the word 'noqa'.

File: coalib/processes/Processing.py [hunk 5]
```diff
@@ -466,6 +468,20 @@ def yield_ignore_ranges(file_dict):
...
                                line_number, 1,
                                end_line, len(file[end_line - 1])))
+                elif 'noqa ' in line:
+                    end_line = min(line_number + 1, len(file))
+                    yield (get_ignore_scope(line, 'noqa '),
+                           SourceRange.from_values(
+                               filename,
+                               line_number, 1,
+                               end_line, len(file[end_line - 1])))
+                elif 'noqa' in line:
```
Comment (RohanVB) [inline coalib/processes/Processing.py]: The second one doesn't have a space after the `noqa`, I needed this to call the 'all bears', otherwise `noqa` would be looking for a string after the keyword and we want `noqa` to do 2 things: 
1)If no bears are specified, `noqa` will run for all bears.
2)If a bear is specified, `noqa` will run for that bear only.

File: coalib/processes/Processing.py [hunk 6]
```diff
@@ -1,10 +1,11 @@
-from itertools import chain
 import multiprocessing
 import os
 import platform
 import queue
 import subprocess
+from itertools import chain
```
Comment (Makman2) [inline coalib/processes/Processing.py]: unrelated change + it breaks alphabetical order (I think we don't distinct between normal `import ...` and `from ... import ...`, just the module name matters)

File: coalib/processes/Processing.py [hunk 7]
```diff
@@ -1,10 +1,11 @@
...
 
+from coalib.collecting import Dependencies
```
Comment (Makman2) [inline coalib/processes/Processing.py]: unused import

File: coalib/processes/Processing.py [hunk 8]
```diff
@@ -443,8 +444,9 @@ def yield_ignore_ranges(file_dict):
         stop_ignoring = False
         for line_number, line in enumerate(file, start=1):
             # Before lowering all lines ever read, first look for the biggest
-            # common substring, case sensitive: I*gnor*e, start i*gnor*ing.
-            if 'gnor' in line:
+            # common substring, case sensitive: I*gnor*e, start i*gnor*ing
+            # and for N*oqa start n*oqa
+            if 'gnor' or 'oqa' in line:
```
Comment (Makman2) [inline coalib/processes/Processing.py]: this does not work sadly, you need to do `'gnor' in line or 'oqa' in line`, otherwise this if always evaluates to `True` (or more exactly, it evaluates to `'gnor'`)

Comment (Makman2) [inline coalib/processes/Processing.py]: or depending on operator priority it could also ignore your changes, as `'gnor' or 'oqa'` evaluates to `'gnor'` which shall be searched in `line`, same behaviour like before

File: coalib/processes/Processing.py (hunk 5 above)
Comment (Makman2) [inline coalib/processes/Processing.py]: if possible could you merge those `if`s together with the `'ignore '` case? you could do something like this:
```python
for ignore_stmt in ['ignore ', 'noqa ', 'noqa']:
    if ignore_stmt in line:
        # construct result accordingly
```

Comment (RohanVB) [inline coalib/processes/Processing.py]: I missed this review, I'll do that, thanks. @Makman2

**Comment by RohanVB** [comment]
Sorry, missed a review. I'm on it.

File: coalib/processes/Processing.py [hunk 9]
```diff
@@ -459,13 +460,15 @@ def yield_ignore_ranges(file_dict):
...
                                    len(file[line_number-1])))
-                elif 'ignore ' in line:
-                    end_line = min(line_number + 1, len(file))
-                    yield (get_ignore_scope(line, 'ignore '),
-                           SourceRange.from_values(
-                               filename,
-                               line_number, 1,
-                               end_line, len(file[end_line - 1])))
+
+                for ignore_stmt in ['ignore ', 'noqa ', 'noqa']:
```
Comment (Makman2) [inline coalib/processes/Processing.py]: I think this should happen in the `else` case to achieve same codeflow like before :)
(also yields slightly performance)

File: coalib/processes/Processing.py [hunk 10]
```diff
@@ -459,13 +460,15 @@ def yield_ignore_ranges(file_dict):
...
+                    if ignore_stmt in line:
+                        end_line = min(line_number + 1, len(file))
+                        yield (get_ignore_scope(line, 'ignore '),
+                               SourceRange.from_values(
+                                   filename,
+                                   line_number, 1,
+                                   end_line, len(file[end_line-1])))
```
Comment (Makman2) [inline coalib/processes/Processing.py]: shall we `break` here when an ignore statement was found?

**Comment by Makman2** [comment]
unack bcee959

File: coalib/processes/Processing.py [hunk 11]
```diff
@@ -459,13 +460,15 @@ def yield_ignore_ranges(file_dict):
...
-                    yield (get_ignore_scope(line, 'ignore '),
-                           SourceRange.from_values(
-                               filename,
-                               line_number, 1,
-                               end_line, len(file[end_line - 1])))
+
+                for ignore_stmt in ['ignore ', 'noqa ', 'noqa']:
+                    if ignore_stmt in line:
+                        end_line = min(line_number + 1, len(file))
+                        yield (get_ignore_scope(line, 'ignore '),
```
Comment (RohanVB) [inline coalib/processes/Processing.py]: Will do. Sorry for missing that.

File: coalib/processes/Processing.py (hunk 10 above)
Comment (RohanVB) [inline coalib/processes/Processing.py]: Sounds good. Will do. :)

File: coalib/processes/Processing.py (hunk 9 above)
Comment (RohanVB) [inline coalib/processes/Processing.py]: adding `elif for ...` :)

**Comment by RohanVB** [comment]
@jayvdb sorry I don't understand, the first test on line 443 is exactly the same as the one on 398 which was already done for ignore. The second test is similar, but returns an empty string which is (->all bears). 
In other words, I just built off the tests which already existed and work fine for ignore. Please elaborate :)

**Comment by RohanVB** [comment]
Ignore the recent push. Fixing it.

File: coalib/processes/Processing.py (hunk 9 above)
Comment (Makman2) [inline coalib/processes/Processing.py]: before it was 
```
                elif 'ignore ' in line:
                    end_line = min(line_number + 1, len(file))
                    yield (get_ignore_scope(line, 'ignore '),
                           SourceRange.from_values(
                               filename,
                               line_number, 1,
                               end_line, len(file[end_line - 1])))
```
so the `elif 'ignore ' in line` was only evaluated if the previous `if ...` failed. I would keep this code flow.

So
> Am I supposed to be doing
> else: for ....

yes :)

> That doesn't work without the code breaking because the for loop requires the if statement for 
> completion. :(

I don't understand that :O

**Comment by Makman2** [comment]
ack 5a0b0c8

**Comment by Makman2** [comment]
@RohanVB rebase and lgtm! :D

**Comment by sils** [comment]
is it marked "WIP" in the title for a reason?

**Comment by RohanVB** [comment]
I did that a few days ago when it was WIP, I guess I forgot to update it.

**Comment by Makman2** [comment]
reack 22ce8f0

**Comment by Makman2** [comment]
aaand we need another rebase :P

**Comment by Mixih** [comment]
This needs to be merged pronto as the keyword as already been merged into doc

**Comment by sils** [comment]
Unmerge the doc or then since the feature is unreleased anyway, rushing
isn't good

**Comment by RohanVB** [comment]
@sils, @makman2 reviewed it above. See: 

> @RohanVB rebase and lgtm! :D

rushing isn't good but I think in this particular case, it's ready?

**Comment by sils** [comment]
No idea, I'm sick, you'll handle :) but even if it's merged it's not
released

**Comment by RohanVB** [comment]
>  I'm sick

Get well soon :)

>  it's not released

good point. Please unmerge https://github.com/coala/documentation/pull/367 until release on this issue. 

CC @Mixih

**Comment by Makman2** [comment]
@rultor merge

**Comment by Mixih** [comment]
eh, it got merged, so I think it's all good ;)

**Comment by RohanVB** [comment]
@Mixih its merged but that would be in the coala --pre version, not the release version - wouldn't it? So the end users are still not gonna have access to it (from what I understand) even though it's merged.

**Comment by Makman2** [comment]
@RohanVB did we release your doc-changes already to stable? Then bad luck or we should backport your commit^^

> its merged but that would be in the coala --pre version, not the release version - wouldn't it? 

yes in the pre version only^^

**Comment by RohanVB** [comment]
doc changes haven't been released to stable, I guess we should be fine. https://github.com/coala/documentation/pull/367

**Comment by Makman2** [comment]
then everything is alright :+1:
